src/appshell/view/editmenucontroller.cpp

```diff
@@ -0,0 +1,144 @@
+//=============================================================================
+//  MuseScore
+//  Music Composition & Notation
+//
+//  Copyright (C) 2021 MuseScore BVBA and others
+//
+//  This program is free software; you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License version 2.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+//=============================================================================
+#include "editmenucontroller.h"
+
+#include "log.h"
+
+using namespace mu::appshell;
+using namespace mu::notation;
+using namespace mu::actions;
+
+void EditMenuController::init()
+{
+    controller()->actionsAvailableChanged().onReceive(this, [this](const std::vector<actions::ActionCode>& actionCodes) {
+        m_actionsReceiveAvailableChanged.send(actionCodes);
+    });
+}
+
+bool EditMenuController::contains(const mu::actions::ActionCode& actionCode) const
+{
+    std::map<ActionCode, AvailableCallback > actions = this->actions();
+    return actions.find(actionCode) != actions.end();
+}
+
+bool EditMenuController::actionAvailable(const mu::actions::ActionCode& actionCode) const
+{
+    std::map<ActionCode, AvailableCallback > actions = this->actions();
+
+    if (actions.find(actionCode) == actions.end()) {
+        return false;
+    }
+
+    return actions[actionCode]();
+}
+
+mu::async::Channel<std::vector<mu::actions::ActionCode> > EditMenuController::actionsAvailableChanged() const
+{
+    return m_actionsReceiveAvailableChanged;
+}
+
+std::map<mu::actions::ActionCode, EditMenuController::AvailableCallback> EditMenuController::actions() const
+{
+    static std::map<ActionCode, AvailableCallback > _actions = {
```

Comment:
Seems that you can simplify this method (same for similar methods in other controllers). It can return just a list of action codes and EditMenuController::actionAvailable will looks like:

actionAvailable(actionCode) {
return controller()->actionAvailable(actionCode);
}

Reply:
Each method can contain additional logic to determine the availability of an action

Reply:
There are currently no checks that contain any additional logic. If we will want to add those checks, we can do this without using any callbacks, just like this:

actionAvailable(actionCode) {
if (actionCode == "codeWithSomeCustomLogic") {
return customCheck()
}

return controller()->actionAvailable(actionCode);
}

Reply:
what happens if half have one logic and the other another? do we list all the code here through if?

Reply:
To be honest, I still doubt that we will ever need a custom check for any code. If such a situation arises, then most likely there will be a small number of such codes. But at the moment we should use std::bind and callbacks for any new action code (in all menu controllers) that we plan to add to the action list. This code is hard to read and it requires many manual operations when you want to add some new action

Reply:
maybe, but it necessary
Next example
Now some menu controllers work with two module controllers(see FileMenuController work with controllers from userscores and notation)

Reply:
You can create two separate lists for those actions (or create an enumerator and use it instead of a large bunch of private methods that are just delegated calls to the controller)